### ISSUE TITLE
Unused font but still subset

### ISSUE THREAD

**Issue by ksesong** [Munter/subfont#51]
When a given font style is font-faced but not used in the body, `subfont` still creates a subset, and includes it in the critical `@font-faces`. It shouldn't create a subset as the font is not used, and strangely enough, the resulting subset is identical to the source file (contains all its glyphs).

This [test repository](https://github.com/ksesong/subfont-duplicate-issue) (which can be processed with `yarn build`) is an example, with the erroneous subset being called `IBM_Plex_Sans-400i-32dd0899cc.woff`.

One thing I also noticed (but which may be unrelated to `subfont`) is that both the italic font and the fallback regular font get downloaded despite not being used.

**Comment by papandreou** [comment]
> When a given font style is font-faced but not used in the body, `subfont` still creates a subset, and includes it in the critical `@font-faces`. It shouldn't create a subset as the font is not used, and strangely enough, the resulting subset is identical to the source file (contains all its glyphs).

That's intentional -- we have to provide all the original variants of the font that we're subsetting in the `__subset`-suffixed font-family. Otherwise the browser will render faux italic/bold if an element utilizing it is injected at runtime. See https://github.com/Munter/subfont/issues/29 for more background.

I've thought about introducing a "purely static"/fallback-less mode where the user attests that text isn't being injected at runtime, so we can avoid all these fallbacks. And #49 could also help avoid the fallbacks in dynamic settings where the complete DOM appears when rendering the page(s) in a headless browser.

> One thing I also noticed (but which may be unrelated to subfont) is that both the italic font and the fallback regular font get downloaded despite not being used.

I'm fairly sure that's also by design. We want the original font to be ready, again in case it's used at runtime in ways `subfont` couldn't decide statically. It should not block the initial rendering, though. Again, we would avoid this in the fallback-less mode.

**Comment by papandreou** [comment]
I'm playing with implementing that "no fallbacks" mode over here: https://github.com/assetgraph/assetgraph/pull/998

It makes the pages in your example repo come out as:

```html
<!DOCTYPE html><html><head>
    <meta charset="UTF-8">
    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2"><style>@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:normal;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");unicode-range:U+20,U+41-5a}</style><script>try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}</script><style>
      body {
        font-family: 'IBM Plex Sans__subset', sans-serif;
        font-style: normal;
        font-weight: 400;
      }
    </style>
  </head>
  <body>
    ABCDEFGHIJKLMNOPQRSTUVWXYZ
  

</body></html>
```

I think that's closer to what you expected to begin with?

**Comment by ksesong** [comment]
Thanks for the links. I understand the motivation, but would it more sensible to have a subset that contains no glyph, so that it fetches the fallbacks later on, rather than at the critical stage? It is a bit strange to subset the regular/bold fonts (to reduce their footprint), and having at the same time a full italic font (which is not even used on the page). (In a way, I could add an invisible italic text with a single glyph so that the italic subset is as small as possible).

The "fallback mode" would work quite well in my situation (and in many static websites). (This wasn't what I expected as I was trying to show an anomaly and I am fine with the fallback fonts being downloaded on a later stage, but this is even better!) This reminds me of [InDesign's font subsetting](https://helpx.adobe.com/indesign/using/pdf-options.html#font_embedding_and_substitution).

**Comment by ksesong** [comment]
I've created another [example repository](https://github.com/ksesong/subfont-missing-issue) to better describe the problem.

There are two files `index-1.html` and `index-2.html`, one of them `index-2.html` contains italics, the other doesn't. After running the script `yarn build`, there is no problem regarding the normal variant, one subset is created for both files in two format (`IBM_Plex_Sans-400-cf9147a291.woff`, `IBM_Plex_Sans-400-a480bd0c27.woff2`). I expect to have one subset for the italic variant as well (in two formats), but I get two subsets, one per `html` file (`index-2.html`: `IBM_Plex_Sans-400i-9d6f8d1af8.woff` and `IBM_Plex_Sans-400i-f1dd0cf5d2.woff2`; `index-1.html` (which doesn't have any italic in its body): `IBM_Plex_Sans-400i-32dd0899cc.woff`). The other strange thing is that the latter file is only in `.woff` format, and as mentioned before it contains all the glyphs from the source version.

Also, this is not directly related to this issue but the fallback styles are still declared in the `<head>`, is  this expected? (the README mentions: *Async load your original @font-face declaring CSS at the bottom of your page, moving it off the critical path*)

**Comment by papandreou** [comment]
> The other strange thing is that the latter file is only in .woff format, and as mentioned before it contains all the glyphs from the source version.

Hmm, it's exactly identical to `IBMPlexSans-Italic.woff` in the source folder. I think the problem is that we accidentally move it to `/subfont/`.

> Also, this is not directly related to this issue but the fallback styles are still declared in the <head>, is this expected? (the README mentions: __Async load your original *@font-face* declaring CSS at the bottom of your page, moving it off the critical path__)

Hmm, you're right. It seems like we never got that implemented for fonts that are originally self-hosted, only Google Web Fonts :man_facepalming:

**Comment by papandreou** [comment]
> Hmm, it's exactly identical to IBMPlexSans-Italic.woff in the source folder. I think the problem is that we accidentally move it to `/subfont/`.

Fixed that in assetgraph 5.8.3.

**Comment by papandreou** [comment]
> Hmm, you're right. It seems like we never got that implemented for fonts that are originally self-hosted, only Google Web Fonts :man_facepalming:

WIP here: https://github.com/assetgraph/assetgraph/pull/999

**Comment by papandreou** [comment]
Now I got `index-2.html` to come out as:

```html
<!DOCTYPE html><html><head>
    <meta charset="UTF-8">
    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2"><style>@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:italic;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400i-9d6f8d1af8.woff) format("woff");unicode-range:U+4e-5a}@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:normal;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");unicode-range:U+20,U+41-5a}</style><script>try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}</script><style>
      body {
        font-family: 'IBM Plex Sans__subset', 'IBM Plex Sans', sans-serif;
        font-style: normal;
        font-weight: 400;
      }
    </style>
  </head>
  <body>
    ABCDEFGHIJKLM<em>NOPQRSTUVWXYZ</em>
  

<script>!function(){var el=document.createElement('link');el.href='/subfont/fallback-423ff769c5.css',el.rel='stylesheet',document.body.appendChild(el)}()</script><noscript><link rel="stylesheet" href="/subfont/fallback-423ff769c5.css"></noscript></body></html>
```

Where `/subfont/fallback-423ff769c5.css` contains:

```css
@font-face{font-family:IBM Plex Sans;font-style:normal;font-weight:400;src:url(/IBMPlexSans-Regular.woff) format("woff")}@font-face{font-family:IBM Plex Sans;font-style:italic;font-weight:400;src:url(/IBMPlexSans-Italic.woff) format("woff")}
```

Or in "no fallbacks" mode:

```html
<!DOCTYPE html><html><head>
    <meta charset="UTF-8">
    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2"><style>@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:italic;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400i-9d6f8d1af8.woff) format("woff");unicode-range:U+4e-5a}@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:normal;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");unicode-range:U+20,U+41-5a}</style><script>try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}</script><style>
      body {
        font-family: 'IBM Plex Sans__subset', sans-serif;
        font-style: normal;
        font-weight: 400;
      }
    </style>
  </head>
  <body>
    ABCDEFGHIJKLM<em>NOPQRSTUVWXYZ</em>
  

</body></html>
```

**Comment by ksesong** [comment]
This looks good to me, how about `index-1.html`? Feel free to close this issue when it is merged.

**Comment by papandreou** [comment]
`index-1.html`:

```html
<!DOCTYPE html><html><head>
    <meta charset="UTF-8">
    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2"><style>@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:normal;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");unicode-range:U+20,U+41-5a}@font-face{font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:italic;font-weight:400;src:url(/IBMPlexSans-Italic.woff) format("woff")}</style><script>try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}</script><style>
      body {
        font-family: 'IBM Plex Sans__subset', 'IBM Plex Sans', sans-serif;
        font-style: normal;
        font-weight: 400;
      }
    </style>
  </head>
  <body>
    ABCDEFGHIJKLMNOPQRSTUVWXYZ
  

<script>!function(){var el=document.createElement('link');el.href='/subfont/fallback-423ff769c5.css',el.rel='stylesheet',document.body.appendChild(el)}()</script><noscript><link rel="stylesheet" href="/subfont/fallback-423ff769c5.css"></noscript></body></html>
```

And without fallbacks:

```html
<!DOCTYPE html><html><head>
    <meta charset="UTF-8">
    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2"><style>@font-face{font-display:swap;font-family:IBM Plex Sans__subset;font-stretch:normal;font-style:normal;font-weight:400;src:url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");unicode-range:U+20,U+41-5a}</style><script>try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}</script><style>
      body {
        font-family: 'IBM Plex Sans__subset', sans-serif;
        font-style: normal;
        font-weight: 400;
      }
    </style>
  </head>
  <body>
    ABCDEFGHIJKLMNOPQRSTUVWXYZ
  

</body></html>
```

**Comment by ksesong** [comment]
I may be wrong in my understanding, but shouldn't the subsets be identical in both pages? (In my very case, I would be using the "no fallbacks" mode, but this is for the sake of completeness).

**Comment by papandreou** [comment]
Not sure exactly what you mean? They look identical to me, here's a diff between the two pages after a manual pretty-printing and rearrangement:

```diff
--- public/index-1.html	2019-04-02 00:23:09.490314982 +0200
+++ public/index-2.html	2019-04-02 00:23:30.502211627 +0200
@@ -1,39 +1,43 @@
 <!DOCTYPE html><html><head>
     <meta charset="UTF-8">
+    <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2">
     <link rel="preload" as="font" type="font/woff2" crossorigin="anonymous" href="/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2">
     <style>
       @font-face {
         font-display: swap;
         font-family: IBM Plex Sans__subset;
         font-stretch: normal;
         font-style: normal;
         font-weight: 400;
         src: url(/subfont/IBM_Plex_Sans-400-246a0cfb85.woff2) format("woff2"),
           url(/subfont/IBM_Plex_Sans-400-cf9147a291.woff) format("woff");
         unicode-range: U+20, U+41-5a;
       }
 
       @font-face {
+        font-display: swap;
         font-family: IBM Plex Sans__subset;
         font-stretch: normal;
         font-style: italic;
         font-weight: 400;
-        src: url(/IBMPlexSans-Italic.woff) format("woff");
+        src: url(/subfont/IBM_Plex_Sans-400i-ff75f98357.woff2) format("woff2"),
+          url(/subfont/IBM_Plex_Sans-400i-9d6f8d1af8.woff) format("woff");
+        unicode-range: U+4e-5a;
       }
     </style>
     <script>
       try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}
     </script>
     <style>
       body {
         font-family: 'IBM Plex Sans__subset', 'IBM Plex Sans', sans-serif;
         font-style: normal;
         font-weight: 400;
       }
     </style>
   </head>
   <body>
-    ABCDEFGHIJKLMNOPQRSTUVWXYZ
+    ABCDEFGHIJKLM<em>NOPQRSTUVWXYZ</em>
   
 
 <script>!function(){var el=document.createElement('link');el.href='/subfont/fallback-423ff769c5.css',el.rel='stylesheet',document.body.appendChild(el)}()</script><noscript><link rel="stylesheet" href="/subfont/fallback-423ff769c5.css"></noscript></body></html>
```

So the differences are that:
* The italic font subset is not preloaded on `index-1.html` (because it isn't used)
* The original italic font is aliased as `IBM Plex Sans__subset` with `font-style: italic`. This is because it's an unused variant that might be utilized at run time (https://github.com/Munter/subfont/issues/29 again).
* No `font-display: swap` for the unused variant on `index-1.html`, where as the italic subset font has it on `index-2.html`. That's intentional, as we actually do want the browser's default behavior in this case.

**Comment by ksesong** [comment]
When it comes to the normal variant, the same subset is used for both pages (despite their bodies having different requirements; `index-2` does not need characters N-Z which are only italic). So I thought the same logic would apply to the italic variant (so exact same subfonts despite the fact that `index-1` body doesn't have any italic character). 

The way I understand it is that the same subsets are used throughout the collection of pages, regardless of the actual content of each individual page, otherwise each page would have its very own subset (which is not the case here, as the normal-style subfont is identical in both pages.)

In the situation the `index-1` page would require some extra characters at runtime, it will be accommodated by the fallback style.

**Comment by papandreou** [comment]
> When it comes to the normal variant, the same subset is used for both pages (despite their bodies having different requirements; index-2 does not need characters N-Z which are only italic)

Side note: The underlying code actually has a "subset per page" mode where the exact characters on each page are used to compute the subsets. It's not currently exposed in `subfont`, though, as we felt that on balance it would be better to always compute site-wide subsets.

> So I thought the same logic would apply to the italic variant (so exact same subfonts despite the fact that index-1 body doesn't have any italic character).

On `index-1.html` we only leave the italic variant there as a fallback, because it's not used according to the static analysis. It wouldn't make sense to use the subset computed for `index-2.html` for that, since we cannot guarantee that unexpected dynamically injected text would fall within that subset. Basically all bets are off in that case, which is why the safest thing is to leave the original font there, so the page will end up looking the same as if subfont hadn't been applied.

The good news is that the italic font variant won't be downloaded by the browser unless that actually happens.

**Comment by ksesong** [comment]
The italic font variant will necessarily be downloaded because of the inline script?
 `try{document.fonts.forEach(function(f){f.family.indexOf('__subset')!==-1&&f.load()})}catch(e){}`

**Comment by papandreou** [comment]
Hmm, yeah, you're right. We'll need to find a strategy to avoid that!

**Comment by papandreou** [comment]
Found a fix: https://github.com/assetgraph/assetgraph/pull/1002

**Comment by ksesong** [comment]
OK :) I have nothing else to suggest, feel free to close this issue when the changes are merged.

**Comment by papandreou** [comment]
All right. Thanks a lot for taking the time to bring up these subjects/bugs in such an actionable way :tada:

**Comment by papandreou** [comment]
Support for the `subfont --no-fallbacks` mode was just released in 3.7.0. With that, everything that has come up here should finally be implemented/fixed :)
